// pico/couchdb_rewrite_url.js
// Generated by CoffeeScript 1.6.3
(function() {
  var couchdb_rewrite_url;

  couchdb_rewrite_url = function(original) {
    var basic, parsed, password, qs, response, url, username, _ref;
    url = require('url');
    qs = require('querystring');
    parsed = url.parse(original);
    response = url.format({
      protocol: parsed.protocol,
      hostname: parsed.hostname,
      port: parsed.port,
      pathname: parsed.pathname
    });
    if (parsed.auth == null) {
      if (parsed.protocol === 'http:' && parsed.hostname === '127.0.0.1' && parsed.port === '5984') {
        return parsed.pathname.substr(1);
      } else {
        return response;
      }
    }
    _ref = parsed.auth.split(/:/), username = _ref[0], password = _ref[1];
    if (username != null) {
      username = qs.unescape(username);
    }
    if (password != null) {
      password = qs.unescape(password);
    }
    if (username == null) {
      username = '';
    }
    if (password == null) {
      password = '';
    }
    basic = new Buffer("" + username + ":" + password);
    response = {
      url: response,
      headers: {
        "Authorization": "Basic " + (basic.toString('base64'))
      }
    };
    return response;
  };

  couchdb_rewrite_url.test = function() {
    var assert;
    assert = require('assert');
    assert.strictEqual(couchdb_rewrite_url('http://127.0.0.1:5984/bob'), 'bob');
    assert.strictEqual(couchdb_rewrite_url('http://example.com:5984/bob'), 'http://example.com:5984/bob');
    assert.deepEqual(couchdb_rewrite_url('http://foo:bar@example.com:5984/bob'), {
      url: 'http://example.com:5984/bob',
      headers: {
        "Authorization": 'Basic Zm9vOmJhcg=='
      }
    });
    return true;
  };

  module.exports = couchdb_rewrite_url;

}).call(this);
